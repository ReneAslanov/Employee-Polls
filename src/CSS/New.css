.new-card-wrapper{
    display: flex;
    flex-direction: column;
    justify-content: center;
    width: 80%;
    margin: 2rem auto;
}

.new-h1{
    margin: 0 auto;
    font-weight: 400;
}

.new-subtitle{
    margin: 0 auto;
    font-weight: 300;
    color: #b3b8c0;
}

.new-label{
    margin: 1rem auto;
}

.new-option-input{
    padding: .7em;
    border: 1px solid #e1e5eb;
    border-radius: 4px;
}

.new-submit{
    margin: 1em auto 0;
    width: 20vw;
    padding: 0.4em 1.5em;
    cursor: pointer;
    background-color: #fff;
    border: 1px solid #e1e5eb;
    text-decoration: none;
    text-align: center;
    color: #b3b8c0;
    background-color: #fff;
    transition: border-color 2s cubic-bezier(0.075, 0.82, 0.165, 1), color 2s cubic-bezier(0.075, 0.82, 0.165, 1), font-size 2s cubic-bezier(0.075, 0.82, 0.165, 1);
}

.new-submit:hover{
    border-color: #000;
    color: #000;
    font-size: 1.2rem;
}
