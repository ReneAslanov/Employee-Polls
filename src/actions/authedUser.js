export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT = "LOGOUT";
export const SET_POLL_OPTION = "SET_POLL_OPTION";

export function setAuthedUser(payload)
{
    return{
        type: SET_AUTHED_USER,
        id: payload.id,
        avatar: payload.avatarURL,
        answers: payload.answers,
        questions: payload.questions
    }
}

export function logoutUser()
{
    return{
        type: LOGOUT
    }
}

export function setPollOption({qid, answer})
{
    return{
        type: SET_POLL_OPTION,
        qid,
        answer
    }
}