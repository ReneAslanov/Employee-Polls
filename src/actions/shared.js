export const RECEIVE_USERS = "RECEIVE_USERS";
export const SET_POLL_ANSWER = "SET_POLL_ANSWER";
export const SET_QUESTION_ARRAY = "SET_QUESTION_ARRAY";

export function setUserAnswer({authedUser, qid, answer})
{
    return{
        type: SET_POLL_ANSWER,
        authedUser,
        qid,
        answer
    }
}

export function setUserQuestionArray({authedUser, qid})
{
  return{
    type: SET_QUESTION_ARRAY,
    authedUser,
    qid
  }
}

export function receiveUsers(payload)
{
    return{
        type: RECEIVE_USERS,
        users: payload
    }
}
