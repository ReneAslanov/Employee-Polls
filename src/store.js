import { legacy_createStore as createStore} from "redux";
import middleware from './middleware';
import reducer from "./reducers/index";

const store = createStore(reducer, middleware);

export default store;

