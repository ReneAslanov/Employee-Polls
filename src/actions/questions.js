import { _saveQuestionAnswer } from "../utils/_DATA.js"
import { SET_POLL_OPTION } from "./authedUser.js";

export const SET_QUESTIONS = "SET_QUESTIONS";
export const CREATE_QUESTION = "CREATE_QUESTION";

export function setQuestions(questions)
{
    return{
        type: SET_QUESTIONS,
        questions : Object.values(questions)
    }
}

export function createQuestion(question)
{
    return{
        type: CREATE_QUESTION,
        payload: question
    }
}

export function handleCreateQuestion(optionOne, optionTwo, author)
{

    return{
        optionOneText: optionOne,
        optionTwoText: optionTwo,
        author
    }
}

export function handleQuestionAnswer({authedUser, qid, answer})
{
    return (dispatch) => {
        return _saveQuestionAnswer({
            authedUser,
            qid,
            answer
        })
        .then(data => {
            return dispatch(setPollAnswer({authedUser, qid, answer}))
        })

    }
}

export function setPollAnswer({qid, answer, authedUser})
{
    return{
        type: SET_POLL_OPTION,
        qid,
        answer,
        authedUser
    }
}