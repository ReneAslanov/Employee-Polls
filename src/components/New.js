import Nav from "./Nav";
import { Link } from "react-router-dom";
import "../CSS/New.css";
import { connect, useDispatch } from "react-redux";
import {useNavigate} from "react-router-dom";
import { createQuestion } from "../actions/questions";
import { _saveQuestion } from "../utils/_DATA";
import { handleCreateQuestion } from "../actions/questions";
import {setUserQuestionArray} from "../actions/shared"
import ErrorPage from "./ErrorPage";

function New(props)
{
    const navigate = useNavigate();
    const dispatch = useDispatch();

    if(props.authedUser === null)
    {
        return(
            <ErrorPage/>
        )
    }

    function dispatchQuestion(e)
    {
        e.preventDefault();

        let optionOne = document.getElementById("option-1").value;
        let optionTwo = document.getElementById("option-2").value;

        if(optionOne === "" || optionTwo === "")
        {
            return alert("please provide both options");
        }
        _saveQuestion(handleCreateQuestion(optionOne, optionTwo, props.authedUser.id))
        .then(actualQuestion => {
            dispatch(setUserQuestionArray({
                authedUser: actualQuestion.author,
                qid: actualQuestion.id
            }));
            return dispatch(createQuestion(actualQuestion))

        })
        setTimeout(navigate, 600, "/home");
    }

    return(
        <div>
            <Nav/>
            <form className="new-card-wrapper">
                <h1 className="new-h1">Would You Rather</h1>
                <h3 className="new-subtitle">Create Your Own Poll</h3>

                <label htmlFor="option-1" className="new-label">First Option</label>
                <input type="text" placeholder="First Option" id="option-1" className="new-option-input" data-testid="option-1"></input>

                <label htmlFor="option-2" className="new-label">Second Option</label>
                <input type="text" placeholder="Second Option" id="option-2" className="new-option-input" data-testid="option-2"></input>

                <Link to="/home" className="new-submit" onClick={dispatchQuestion}>Submit</Link>
            </form>
        </div>
    )
}

function mapStateToProps({authedUser})
{
    if(authedUser === null)
    {
        return{
            authedUser: null
        }
    }

    return{
        authedUser,
    }
}

export default connect(mapStateToProps)(New);