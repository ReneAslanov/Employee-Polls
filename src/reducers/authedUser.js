import { SET_AUTHED_USER, SET_POLL_OPTION } from "../actions/authedUser";
import { LOGOUT } from "../actions/authedUser";
import { CREATE_QUESTION } from "../actions/questions";

export default function authedUser(state = null, action)
{
    switch(action.type)
    {
        case SET_AUTHED_USER:
            return{
               id: action.id,
               avatar: action.avatar,
               answers: action.answers,
               questions: action.questions
            };

        case LOGOUT:
            return null

        case CREATE_QUESTION:
            return{
                ...state,
                questions: state.questions.concat(action.payload.id)
            }

        case SET_POLL_OPTION:

            const {qid, answer} = action;
            return{
                ...state,
                answers: {
                    ...state.answers,
                    [qid]: answer
                }
            }

        default:
            return state;
    }
}