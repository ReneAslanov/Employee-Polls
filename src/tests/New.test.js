import New from "../components/New";
import store from "../store";
import {setAuthedUser} from "../actions/authedUser";
import { MemoryRouter } from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";

describe("New", () => {
    it("should render the component properly and the input fields should have proper values when changed", () => {
        const testAuthedUser = {
            id: "sarahedo",
            avatarURL: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU",
            answers: {
                "8xf0y6ziyjabvozdd253nd": 'optionOne',
                "6ni6ok3ym7mf1p33lnez": 'optionOne',
                "am8ehyc8byjqgar0jgpub9": 'optionTwo',
                "loxhs1bqm25b708cmbf3g": 'optionTwo'
              },
            questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
        }

        store.dispatch(setAuthedUser(testAuthedUser));

        const view = render(
            <MemoryRouter initialEntries={["/home"]}>
                <Provider store={store}>
                    <New/>
                </Provider>
            </MemoryRouter>
        )

        expect(view).toMatchSnapshot();
        expect(screen.getByText("Would You Rather")).toBeInTheDocument();

        const inputOne = screen.getByTestId("option-1");
        const inputTwo = screen.getByTestId("option-2");

        fireEvent.change(inputOne, {target: {value: "test option one"}});
        fireEvent.change(inputTwo, {target: {value: "test option two"}});

        expect(inputOne.value).toEqual("test option one");
        expect(inputTwo.value).toEqual("test option two");
    })

})