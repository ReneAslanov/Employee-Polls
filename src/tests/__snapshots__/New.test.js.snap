// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New should render the component properly and the input fields should have proper values when changed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="nav-wrapper"
        >
          <div
            class="nav-routes"
          >
            <ul
              class="nav-list"
            >
              <a
                href="/home"
              >
                Home
              </a>
              <a
                href="/leaderboard"
              >
                Leaderboard
              </a>
              <a
                href="/add"
              >
                New
              </a>
            </ul>
          </div>
          <div
            class="nav-user"
          >
            <ul
              class="nav-list"
            >
              <div
                class="nav-img-wrapper"
              >
                <img
                  alt="just an avatar"
                  class="nav-avatar"
                  src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU"
                />
              </div>
              <p
                class="nav-userName"
              >
                sarahedo
              </p>
              <a
                href="/"
              >
                Logout
              </a>
            </ul>
          </div>
        </div>
        <form
          class="new-card-wrapper"
        >
          <h1
            class="new-h1"
          >
            Would You Rather
          </h1>
          <h3
            class="new-subtitle"
          >
            Create Your Own Poll
          </h3>
          <label
            class="new-label"
            for="option-1"
          >
            First Option
          </label>
          <input
            class="new-option-input"
            data-testid="option-1"
            id="option-1"
            placeholder="First Option"
            type="text"
          />
          <label
            class="new-label"
            for="option-2"
          >
            Second Option
          </label>
          <input
            class="new-option-input"
            data-testid="option-2"
            id="option-2"
            placeholder="Second Option"
            type="text"
          />
          <a
            class="new-submit"
            href="/home"
          >
            Submit
          </a>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="nav-wrapper"
      >
        <div
          class="nav-routes"
        >
          <ul
            class="nav-list"
          >
            <a
              href="/home"
            >
              Home
            </a>
            <a
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              href="/add"
            >
              New
            </a>
          </ul>
        </div>
        <div
          class="nav-user"
        >
          <ul
            class="nav-list"
          >
            <div
              class="nav-img-wrapper"
            >
              <img
                alt="just an avatar"
                class="nav-avatar"
                src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU"
              />
            </div>
            <p
              class="nav-userName"
            >
              sarahedo
            </p>
            <a
              href="/"
            >
              Logout
            </a>
          </ul>
        </div>
      </div>
      <form
        class="new-card-wrapper"
      >
        <h1
          class="new-h1"
        >
          Would You Rather
        </h1>
        <h3
          class="new-subtitle"
        >
          Create Your Own Poll
        </h3>
        <label
          class="new-label"
          for="option-1"
        >
          First Option
        </label>
        <input
          class="new-option-input"
          data-testid="option-1"
          id="option-1"
          placeholder="First Option"
          type="text"
        />
        <label
          class="new-label"
          for="option-2"
        >
          Second Option
        </label>
        <input
          class="new-option-input"
          data-testid="option-2"
          id="option-2"
          placeholder="Second Option"
          type="text"
        />
        <a
          class="new-submit"
          href="/home"
        >
          Submit
        </a>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
